content/docs/hoon/guides/sail.md

```diff
@@ -0,0 +1,495 @@
+  ; {txt}
+  {<our>} {<now>} {<`@ux`(end 6 eny)>}
+==
+```
+
+Produces:
+
+```
+<article>
+  <b>33138</b> bananas
+  <p>~zod ~2022.2.21..09.54.21..5b63 0x9827.99c7.06f4.8ef9</p>
+</article>
+```
+
+## A note on CSS
+
+The CSS for a page is usually quite large. The typical approach is to include a
+separate arm in your agent (`++style` or the like) and write out the CSS in a
+fenced cord block. You can then call `++trip` on it and include it in a style
+tag. For example:
+
+```hoon
+++  style
+  ^~
+  %-  trip
+  '''
+  main {
+    width: 100%;
+    color: red;
+  }
+  header {
+    color: blue;
+    font-family: monospace;
+  }
+  '''
+```
+
+And then your style tag might look like:
+
+```hoon
+;style: {style}
+```
+
+A cord is used rather than a tape so you don't need to escape double-quotes. The
+[ketsig](/docs/hoon/reference/rune/ket#-ketsig) (`^~`) rune means `++trip` will
+be run at compile time rather than call time.
+
+## Types and marks
+
+So far we've shown rendered HTML for demonstrative purposes, but Sail syntax
+doesn't directly produce HTML text. Instead, it produces a
+[$manx](/docs/hoon/reference/stdlib/5e#manx). This is a Hoon type used to
+represent an XML hierarchical structure with a single root node. There are six
+XML-related types defined in the standard library:
+
+```hoon
++$  mane  $@(@tas [@tas @tas])                    ::  XML name+space
++$  manx  $~([[%$ ~] ~] [g=marx c=marl])          ::  dynamic XML node
++$  marl  (list manx)                             ::  XML node list
++$  mars  [t=[n=%$ a=[i=[n=%$ v=tape] t=~]] c=~]  ::  XML cdata
++$  mart  (list [n=mane v=tape])                  ::  XML attributes
++$  marx  $~([%$ ~] [n=mane a=mart])              ::  dynamic XML tag
+```
+
+More information about these can be found in [section 5e of the standard library
+reference](/docs/hoon/reference/stdlib/5e).
+
+You don't need to understand these types in order to write Sail. The main thing
+to note is that a `$manx` is a node (a single tag) and its contents is a
+[$marl](/docs/hoon/reference/stdlib/5e#marl), which is just a `(list manx)`.
+
+### Rendering
+
+A `$manx` can be rendered as HTML in a tape with the `++en-xml:html` function in
+`zuse.hoon`. For example:
+
+```
+> ;p: foobar
+[[%p ~] [[%$ [%$ "foobar"] ~] ~] ~]
+
+> =x ;p: foobar
+
+> (en-xml:html x)
+"<p>foobar</p>"
+
+> (crip (en-xml:html x))
+'<p>foobar</p>'
+```
+
+You may wish to do this in your agent directly, but you can also leave mark
+conversion to Eyre.
+
+### Sanitization
+
+The `++en-xml:html` function will sanitize the contents of both attributes and
+elements, converting characters such as `>` to HTML entities. For example:
+
+```
+> =z ;p(class "\"><script src=\"example.com/xxx.js"): <h1>FOO</h1>
+
+> (crip (en-xml:html z))
+'<p class="&quot;&gt;&lt;script src=&quot;example.com/xxx.js">&lt;h1&gt;FOO&lt;/h1&gt;</p>'
+```
+
+### Marks
+
+The mark for a raw `$manx` is a `%hymn`. This mark is included in the `%base`
```

Comment:
I think a little summary table of manx/marl/hymn etc. could be helpful.

Reply:
Wouldn't it make more sense to just go ahead and change the mark `%hymn` to a name that makes sense like `%manx` or `%xml` lol?

Reply:
I'll redo this sections cos it's a bit inaccurate. Technically a `%hymn` is an HTML document, and an `%elem` is any xml node. They're both manxes in terms of their type but `%hymn` is intended to have a proper html > head & body structure while an `%elem` is not. There is already an `%xml` mark, along with `%htm`, `%html` `%x-htm`, `%x-htm-elem`. The `%hymn` mark is used throughout these in terms of mark conversions as well as in other source code so can't really change without disentangling a whole bunch of mark conversion pathways and changing other stuff